spec/abci++/v3.md

```diff
@@ -0,0 +1,202 @@
+# Tendermint v3 Markdown pseudocode
+
+This is a single-threaded implementation of ABCI++,
+with an optimization for the ProcessProposal phase.
+Namely, processing of the header and the block data is separated into two different functions.
+
+### Initialization
+
+```
+h_p = 0
+round_p = 0
+step_p is one of {propose, prevote, precommit}
+decision_p = nil
+lockedValue_p = nil
+validValue_p = nil
+validRound_p = -1
+```
+
+### StartRound(round)
+
+```go
+function startRound(round) {
+    round_p ← round
+    step_p ← propose
+    if proposer(h_p, round_p) = p {
+        if validValue_p != nil {
+            proposal ← validValue_p
+        }
+        else {
+            txdata ← mempool.GetBlock()
```

Comment:
Speaking about the mempool, I think it would be great if the mempool connection was also mentioned soemwhere (not necessarily in the pseudo code and just for the sake of completeness).